internal/cmdrender/executor.go

```diff
@@ -540,3 +567,150 @@ func pruneResources(hctx *hydrationContext) error {
+}
+
+// nameMatch returns true if the resource name matches input selection criteria
+func nameMatch(node *yaml.RNode, selector kptfilev1.Selector) bool {
+	return selector.Name == "" || selector.Name == node.GetName()
+}
+
+// namespaceMatch returns true if the resource namespace matches input selection criteria
+func namespaceMatch(node *yaml.RNode, selector kptfilev1.Selector) bool {
+	return selector.Namespace == "" || selector.Namespace == node.GetNamespace()
+}
+
+// kindMatch returns true if the resource kind matches input selection criteria
+func kindMatch(node *yaml.RNode, selector kptfilev1.Selector) bool {
+	return selector.Kind == "" || selector.Kind == node.GetKind()
+}
+
+// apiVersionMatch returns true if the resource apiVersion matches input selection criteria
+func apiVersionMatch(node *yaml.RNode, selector kptfilev1.Selector) bool {
+	return selector.APIVersion == "" || selector.APIVersion == node.GetApiVersion()
+}
+
+// packagePathMatch returns true if the package path of resource matches input selection criteria
+func packagePathMatch(node *yaml.RNode, selector kptfilev1.Selector, rootPackagePath string) (bool, error) {
+	resourcePkgPath, err := pkg.GetPkgPathAnnotation(node)
+	if err != nil {
+		return false, err
+	}
+	return selector.PackagePath == "" || filepath.Join(rootPackagePath, selector.PackagePath) == resourcePkgPath, nil
+}
+
+// setResourceIds adds resource-id annotation to each input resource
+func setResourceIds(input []*yaml.RNode) error {
+	id := 0
+	for i := range input {
+		idStr := fmt.Sprintf("%v", id)
+		err := input[i].PipeE(yaml.SetAnnotation(resourceIDAnnotation, idStr))
+		if err != nil {
+			return err
+		}
+		id++
+	}
+	return nil
+}
+
+// mergeWithInput merges the transformed output with input resources
+// input: all input resources, selectedInput: selected input resources
+// output: output resources as the result of function on selectedInput resources
+// for input: A,B,C,D; selectedInput: A,B; output: A,E(A transformed, B Deleted, E Added)
+// the result should be A,C,D,E
+// resources are identified by resource-id annotation
+func mergeWithInput(output, selectedInput, input []*yaml.RNode) []*yaml.RNode {
+	var result []*yaml.RNode
+	for i := range input {
+		if !presentIn(input[i], selectedInput) {
+			// this resource is untouched
+			result = append(result, input[i])
+			continue
+		}
+
+		if presentIn(input[i], selectedInput) && !presentIn(input[i], output) {
+			// this resource is deleted
+			continue
+		}
+
+		if presentIn(input[i], selectedInput) && presentIn(input[i], output) {
+			// this resource modified by function, so replace it from the output
+			result = append(result, nodeWithResourceID(input[i].GetAnnotations()[resourceIDAnnotation], output))
+		}
+	}
+
+	// add function generated resources to the result
+	for i := range output {
+		if output[i].GetAnnotations()[resourceIDAnnotation] == "" {
+			result = append(result, output[i])
+		}
+	}
+
+	return result
+}
```

Comment:
There are many PresentIn calls (O(n)) so we should keep an eye on performance.

Reply:
Good catch. Simplified the logic. There will be only 2 calls.